components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/StreamingListener.java

```diff
@@ -0,0 +1,163 @@
+                builder.discoverPrefix(discoverPrefix);
+            }
+
+            if (StringUtils.isNotEmpty(maxPingsOut)) {
+                builder.maxPingsOut(Integer.parseInt(maxPingsOut));
+            }
+
+            if (StringUtils.isNotEmpty(pingInterval)) {
+                builder.pingInterval(Duration.ofSeconds(Integer.parseInt(pingInterval)));
+            }
+
+            if (Boolean.parseBoolean(traceConnection)) {
+                builder.traceConnection();
+            }
+
+            StreamingConnectionFactory connectionFactory = new StreamingConnectionFactory(builder.build());
+            connection = connectionFactory.createConnection();
+        }
+        return true;
+    }
+
+    @Override public void consumeMessage(String sequenceName)
+            throws InterruptedException, IOException, TimeoutException {
+        CountDownLatch latch = new CountDownLatch(1);
+
+        SubscriptionOptions.Builder subscriptionOptions = new SubscriptionOptions.Builder();
+
+        String durableName = natsProperties.getProperty(NatsConstants.NATS_STREAMING_DURABLE_NAME);
+        String queueGroup = natsProperties.getProperty(NatsConstants.NATS_STREAMING_QUEUE_GROUP);
+        boolean isManualAck = Boolean.parseBoolean(natsProperties.getProperty(NatsConstants.NATS_STREAMING_MANUAL_ACK));
+        String ackWait = natsProperties.getProperty(NatsConstants.NATS_STREAMING_ACK_WAIT);
+        String maxInFlight = natsProperties.getProperty(NatsConstants.NATS_STREAMING_MAX_IN_FLIGHT);
+        String subscriptionTimeout = natsProperties.getProperty(NatsConstants.NATS_STREAMING_SUBSCRIPTION_TIMEOUT);
+        String dispatcher = natsProperties.getProperty(NatsConstants.NATS_STREAMING_DISPATCHER);
+
+        if (StringUtils.isNotEmpty(durableName)) {
+            subscriptionOptions.durableName(durableName);
+        }
+
+        if (isManualAck) {
+            subscriptionOptions.manualAcks();
+        }
+
+        if (StringUtils.isNotEmpty(ackWait)) {
+            subscriptionOptions.ackWait(Duration.ofSeconds(Integer.parseInt(ackWait)));
+        }
+
+        if (StringUtils.isNotEmpty(maxInFlight)) {
+            subscriptionOptions.maxInFlight(Integer.parseInt(maxInFlight));
+        }
+
+        if (StringUtils.isNotEmpty(subscriptionTimeout)) {
+            subscriptionOptions.subscriptionTimeout(Duration.ofSeconds(Integer.parseInt(subscriptionTimeout)));
+        }
+
+        if (StringUtils.isNotEmpty(dispatcher)) {
+            subscriptionOptions.dispatcher(dispatcher);
+        }
+
+        if (StringUtils.isNotEmpty(queueGroup)) {
+            connection.subscribe(subject, queueGroup, natsMessage -> {
+                String message = new String(natsMessage.getData());
+                log.info("Message Received to NATS Inbound EP: " + message);
+                injectHandler.invoke(message.getBytes(), sequenceName, null, null);
+                latch.countDown();
+                manuallyAck(isManualAck, natsMessage);
```

Comment:
Someone putting manually ack would mean to acknowledge the message only if the mediation of the message is successful. In that case, this is not the correct place to acknowledge the message manually. We need to do it after mediation is successful.

Reply:
Let's modify the logic accordingly. Also make sure to document our interpretation of this parameter.